taichi/backends/cuda/codegen_cuda.cpp

```diff
@@ -240,6 +240,48 @@ class CodeGenLLVMCUDA : public CodeGenLLVM {
 #undef UNARY_STD
   }
 
+  // Not all reduction statements can be optimized.
+  // If the operation cannot be optimized, this function returns nullptr.
+  llvm::Value *optimized_reduction(AtomicOpStmt *stmt) {
+    if (!stmt->is_reduction) {
+      return nullptr;
+    }
+    if (!stmt->val->ret_type->is<PrimitiveType>()) {
+      return nullptr;
+    }
+    PrimitiveTypeID prim_type =
+        stmt->val->ret_type->cast<PrimitiveType>()->type;
+
+    std::unordered_map<PrimitiveTypeID,
+                       std::unordered_map<AtomicOpType, std::string>>
+        fast_reductions;
+
+    fast_reductions[PrimitiveTypeID::i32][AtomicOpType::add] = "reduce_add_i32";
+    fast_reductions[PrimitiveTypeID::f32][AtomicOpType::add] = "reduce_add_f32";
+    fast_reductions[PrimitiveTypeID::i32][AtomicOpType::min] = "reduce_min_i32";
+    fast_reductions[PrimitiveTypeID::f32][AtomicOpType::min] = "reduce_min_f32";
+    fast_reductions[PrimitiveTypeID::i32][AtomicOpType::max] = "reduce_max_i32";
+    fast_reductions[PrimitiveTypeID::f32][AtomicOpType::max] = "reduce_max_f32";
+
+    fast_reductions[PrimitiveTypeID::i32][AtomicOpType::bit_and] =
+        "reduce_and_i32";
+    fast_reductions[PrimitiveTypeID::i32][AtomicOpType::bit_or] =
+        "reduce_or_i32";
+    fast_reductions[PrimitiveTypeID::i32][AtomicOpType::bit_xor] =
+        "reduce_xor_i32";
+
+    AtomicOpType op = stmt->op_type;
+    if (fast_reductions.find(prim_type) == fast_reductions.end()) {
+      return nullptr;
+    }
+    if (fast_reductions.at(prim_type).find(op) ==
```

Comment:
Similarly, if `op` is not in {add, min, max}, it will be a bug in our code. So we can just remove this?

Reply:
Yes, I think we can replace it with a TI_ASSERT for now.

Reply:
Thanks. But it would still be a bug in our code if we generate a pair of `f32/xor`

Reply:
Shall I do a `TI_ASSERT(fast_reductions.at(prim_type).find(op) !=fast_reductions.at(prim_type).end())` then?

Reply:
Sounds good!